controllers/index.js

```diff
@@ -10,10 +10,47 @@ router.get('/', async (req, res, next) => {
     // Fetch templates
     const {templates} = await model.list(0, null);
     const filterOptions = {
-      sort: req.query.sort || 'views',
-      templateTypes: (req.query.templateTypes || 'all').split(','),
-      categories: (req.query.categories || 'all').split(',')
+      sort: (req.query.sort) ? req.query.sort.split(',') : ['views'],
+      templateTypes: (req.query.templateTypes) ? req.query.templateTypes.split(',') : ['all'],
+      categories: (req.query.categories) ? req.query.categories.split(',') : ['all']
     };
+    
+    // Sanitizing refining values
+    // Template Types
+    filterOptions.templateTypes.forEach(function(e,i){
+        if(Object.keys(enums.allowedFilterValues.templateTypes).indexOf(e)===-1) filterOptions.templateTypes.splice(i,1);
+    });
+    // User may remove the filter value manually and key will still exist, set default value
+    if(filterOptions.templateTypes.length===0) filterOptions.templateTypes = ['all'];
+
+    // Sort Types
+    filterOptions.sort.forEach(function(e,i){
+        if(Object.keys(enums.allowedFilterValues.sort).indexOf(e)===-1) filterOptions.sort.splice(i,1);
+    });
+    // User may remove the filter value manually and key will still exist, set default value
+    if(filterOptions.sort.length===0) filterOptions.sort = ['views'];
+
+    // Categories
+    filterOptions.categories.forEach(function(e,i){
+        if(Object.keys(enums.allowedFilterValues.categories).indexOf(e)===-1) filterOptions.categories.splice(i,1);
+    });
+    // User may remove the filter value manually and key will still exist, set default value
+    if(filterOptions.categories.length===0) filterOptions.categories = ['all'];      
+
+      
+   /*   
+    if(Object.keys(enums.allowedFilterValues.sort).indexOf(filterOptions.sort)===-1){
+        filterOptions.sort = 'views';
+    }
+      
+    filterOptions.templateTypes.forEach(function(e){
+        if(Object.keys(enums.filterValues.sort).indexOf(filterOptions.sort)===-1){
+            
+        }
+    });    
+    */
+    // CleanUp Filters (Remove non-existing ones)
+    
 
```

Comment:
All this new sanitation should be done `gtmTplParser.filterAndSort` - no need to separate sanitation from the actual filtering and sorting - makes it really difficult to test.

Reply:
@sahava I didn't add it to filterAndSort because We need to pass the info to the template and to the dataLayer.

Reply:
That's fine, but then there should be a method like `sanitize` so that you can do 

```
const filterOptions = {
  sort: gtmTemplateParser.sanitize(req.query.sort),
  templateTypes: gtmTempalteParser.sanitize(req.query.templateTypes),
  categories: gtmTemplateParser.sanitize(req.query.categories)
};
```

and that `sanitize` method would set the default value to ['all']/['views'] and then do the check against the enum. 

The problem with the current setup is that you've added a lot of parsing logic to the controller, when the controller should just take an input (the request), parse it with middleware and helpers, and output a response. It's very difficult to test side effects in a controller, we need to be able to do atomic tests against each helper method, which is why they should be in the helper library.

You also have basically the same JavaScript written three times (the check against the enum), when it should be just one written method called with different parameters.

Reply:
moved all to sanitize function within templates helper